File: native/android/core-model/src/main/kotlin/com/fireball/nativeapp/core/model/ScrobbleRules.kt
package com.fireball.nativeapp.core.model

/**
 * Client-side scrobble timing (ListenBrainz + Last.fm).
 *
 * Fires when **either** [percent] of the track or [maxSeconds] elapsed is reached first.
 * Tracks shorter than [minTrackSeconds] are never scrobbled (Last.fm minimum; common LB clients).
 *
 * ListenBrainz documents submission at half the track or 4 minutes (whichever is lower);
 * Fireball defaults (25% / 40s) are more conservative and user-tunable in settings.
 */
object ScrobbleRules {
    fun shouldScrobble(
        positionMs: Long,
        durationMs: Long,
        percent: Int,
        maxSeconds: Int,
        minTrackSeconds: Int = DEFAULT_MIN_TRACK_SECONDS,
    ): Boolean {
        if (durationMs <= 0L) return false
        val minMs = minTrackSeconds.coerceAtLeast(1) * 1000L
        if (durationMs < minMs) return false
        val pct = percent.coerceIn(0, 100)
        val maxSec = maxSeconds.coerceAtLeast(0)
        val byPercent = pct > 0 && positionMs >= (durationMs * pct / 100.0).toLong()
        val byMaxSec = maxSec > 0 && positionMs >= maxSec * 1000L
        return byPercent || byMaxSec
    }

    const val DEFAULT_MIN_TRACK_SECONDS = 30
    const val DEFAULT_PERCENT = 25
    const val DEFAULT_MAX_SECONDS = 40
}

File: native/ios/FireballNative/Core/ScrobbleRules.swift
import Foundation

/// Client-side scrobble timing (ListenBrainz + Last.fm).
/// Fires when **either** percent of the track or max seconds is reached first.
/// Tracks shorter than `minTrackSeconds` are never scrobbled.
enum ScrobbleRules {
    static let defaultMinTrackSeconds = 30
    static let defaultPercent = 25
    static let defaultMaxSeconds = 40

    static func shouldScrobble(
        positionMs: Int64,
        durationMs: Int64,
        percent: Int,
        maxSeconds: Int,
        minTrackSeconds: Int = defaultMinTrackSeconds
    ) -> Bool {
        guard durationMs > 0 else { return false }
        let minMs = Int64(max(1, minTrackSeconds)) * 1000
        guard durationMs >= minMs else { return false }
        let pct = min(100, max(0, percent))
        let maxSec = max(0, maxSeconds)
        let byPercent = pct > 0 && positionMs >= durationMs * Int64(pct) / 100
        let byMaxSec = maxSec > 0 && positionMs >= Int64(maxSec) * 1000
        return byPercent || byMaxSec
    }

    static func shouldScrobble(
        positionSeconds: Double,
        durationSeconds: Double,
        percent: Int,
        maxSeconds: Int,
        minTrackSeconds: Int = defaultMinTrackSeconds
    ) -> Bool {
        guard durationSeconds > 0 else { return false }
        return shouldScrobble(
            positionMs: Int64(positionSeconds * 1000),
            durationMs: Int64(durationSeconds * 1000),
            percent: percent,
            maxSeconds: maxSeconds,
            minTrackSeconds: minTrackSeconds
        )
    }
}
